ref.go

```diff
@@ -111,15 +119,18 @@ func (e *Encoder) checkRefMap(v reflect.Value) (int, bool) {
 	}
 
 	if elem, ok := e.refMap[addr]; ok {
-		// the array addr is equal to the first elem, which must ignore
 		if elem.kind == kind {
-			return elem.index, ok
+			if elem.kind != reflect.Struct {
+				return elem.index, ok
+			} else if elem.tp == tp {
```

Comment:
the `else if` means the `elem` is  a struct,  when will its type being not equal if it's kind is equal? pls add more comment about this logic, tks

Reply:
- Firstly kind must be equal.
- Secondly there are two situations:
   - if kind is not `reflect.Struct`, just return `elem.index, ok`
   - if kind is `reflet.Struct`, we must ensure the types must equal 

The reason that why we need to use type information to distinguish `struct` is that different empty struct may share the same address and kind.

```go
package main

import (
    "fmt"
    "reflect"
    "unsafe"
)

type A0 struct{}

func (*A0) JavaClassName() string {
    return "com.caucho.hessian.test.A0"
}

type A1 struct{}

func (*A1) JavaClassName() string {
    return "com.caucho.hessian.test.A1"
}

func main() {
    a0 := new(A0)
    a1 := new(A1)
    v1 := reflect.ValueOf(a0)
    v2 := reflect.ValueOf(a1)
    fmt.Println(unsafe.Pointer(a0))  // 0x5761c8
    fmt.Println(unsafe.Pointer(a1))  // 0x5761c8
    fmt.Println(unsafe.Sizeof(a0))   // 8
    fmt.Println(unsafe.Sizeof(a1))   // 8
    fmt.Println(v1.Elem().Kind())    // struct
    fmt.Println(v2.Elem().Kind())    // struct
    fmt.Println(v1.Elem().Type())    // main.A0
    fmt.Println(v2.Elem().Type())    // main.A1
}
```

Futher more, if we don't use the type information to distinguish `struct`, the code below will run wrongly

```go
func TestEmptyStruct(t *testing.T) {
	expected := []interface{}{new(A0), new(A1)}
	
	e := NewEncoder()
	e.Encode(expected)
	
	d := NewDecoder(e.Buffer())
	result, _ := d.Decode()
	
	if !reflect.DeepEqual(expected, result) {
		t.Errorf("encoding wrongly")
	}
}
```

Because `new(A0)` and `new(A1)` share the same address and kind and encoder think they are the same,  then `result` will be `[]interface{}{new(A0), new(A0)}`.

Reply:
got it, this situation seems only occurs for empty struct, the key description above can be added in code comment.

Reply:
@wongoo I have add comments:

https://github.com/dubbogo/hessian2/blob/ad9cb8199f7324a978de9d46309d13f92ae5b070/ref.go#L31-L33

https://github.com/dubbogo/hessian2/blob/ad9cb8199f7324a978de9d46309d13f92ae5b070/ref.go#L124-L131